src/win_cmds.c

```diff
@@ -86,44 +90,50 @@ process_cmd(void)
 	return 0;
 }
 
+static int
+add_key_to_buffer(int c, char *buf, int *pi) {
+	if ((win_flags & F_KEY_ESC) || c == '\n') { /* cancel/confirm */
+		if (c == '\n')
+			return 1; /* confirm (enter) */
+		return 2; /* cancel (escape) */
+	} else {
+		if (c == KEY_BACKSPACE) {
+			if (--(*pi) < 0)
+				*pi = 0;
+		} else if (*pi < MAX_BUF-1) {
+			buf[(*pi)++] = c;
+		}
+		buf[*pi] = '\0';
+	}
+
+	return 0;
+}
+
 static void
 cmds_handle_key(int c)
 {
-	static int i = 0;
-	int code = 0;
-	char *inp;
-
-	if (win_flags & F_WCMDS_SRCH)
-		inp = wc.srch;
-	else
-		inp = wc.cmd;
-
-	if (!visible)
-		return;
-
-	if (c == '\n') { /* confirm */
-		if (win_flags & F_WCMDS_CMDS)
-			code = process_cmd();
-
-		if (code) {
-			return;
-		}
+	static int si = 0; /* search index */
+	static int ci = 0; /* command index */
+	int ret = 0;
+
+	switch (win_flags & TUI_WCMDS_MASK) {
+	case F_WCMDS_SRCH: /* search mode */
+		ret = add_key_to_buffer(c, wc.srch, &si);
+		if (ret == 2)
+			wc.srch[si=0] = '\0';
+		break;
+	case F_WCMDS_CMDS: /* command mode */
+		ret = add_key_to_buffer(c, wc.cmd, &ci);
+		if (ret == 1)
+			process_cmd();
+		if (ret == 1 || ret == 2)
+			wc.cmd[ci=0] = '\0';
+		break;
 	}
 
-	if ((win_flags & F_KEY_ESC) || c == '\n') { /* cancel/finish */
-		if (win_flags & (F_KEY_ESC | F_WCMDS_CMDS))
-			inp[i=0] = '\0';
+	if (ret) {
 		win_flags &= ~TUI_WCMDS_MASK;
-		windows_visible(WIN_CMDS, 0);
```

Comment:
this function is not needed anymore... as the draw function doesn't show anything if it's not in CMDS or SRCH mode.
Should we delete 'cmds_visible'?